src/image/loading_displaying.js

```diff
@@ -158,6 +159,387 @@ p5.prototype.loadImage = function(path, successCallback, failureCallback) {
+ * 'frames'. By default it's 'seconds'.
+ *
+ * @example
+ * <div>
+ * <code>
+ * function setup() {
+ *   createCanvas(100, 100);
+ * }
+ *
+ * function draw() {
+ *   colorMode(RGB);
+ *   background(30);
+ *
+ *   // create a bunch of circles that move in... circles!
+ *   for (let i = 0; i < 10; i++) {
+ *     let opacity = map(i, 0, 10, 0, 255);
+ *     noStroke();
+ *     fill(230, 250, 90, opacity);
+ *     circle(
+ *       30 * sin(frameCount / (30 - i)) + width / 2,
+ *       30 * cos(frameCount / (30 - i)) + height / 2,
+ *       10
+ *     );
+ *   }
+ * }
+ *
+ * // you can put it in the mousePressed function,
+ * // or keyPressed for example
+ * function keyPressed() {
+ *   // this will download the first 5 seconds of the animation!
+ *   if (key === 's') {
+ *     saveGif('mySketch', 5);
+ *   }
+ * }
+ * </code>
+ * </div>
+ *
+ * @alt
+ * animation of a circle moving smoothly diagonally
+ */
+p5.prototype.saveGif = async function(
+  fileName,
+  duration,
+  options = { delay: 0, units: 'seconds' }
+) {
+  // validate parameters
+  if (typeof fileName !== 'string') {
+    throw TypeError('fileName parameter must be a string');
+  }
+  if (typeof duration !== 'number') {
+    throw TypeError('Duration parameter must be a number');
+  }
+  // if arguments in the options object are not correct, cancel operation
+  if (typeof options.delay !== 'number') {
+    throw TypeError('Delay parameter must be a number');
+  }
+  // if units is not seconds nor frames, throw error
+  if (options.units !== 'seconds' && options.units !== 'frames') {
+    throw TypeError('Units parameter must be either "frames" or "seconds"');
+  }
+
+  // extract variables for more comfortable use
+  let units = options.units;
+  let delay = options.delay;
+
+  //   console.log(options);
+
+  // get the project's framerate
+  let _frameRate = this._targetFrameRate;
+  // if it is undefined or some non useful value, assume it's 60
+  if (_frameRate === Infinity || _frameRate === undefined || _frameRate === 0) {
+    _frameRate = 60;
+  }
+
+  // calculate frame delay based on frameRate
+
+  // this delay has nothing to do with the
+  // delay in options, but rather is the delay
+  // we have to specify to the gif encoder between frames.
+  let gifFrameDelay = 1 / _frameRate * 1000;
+
+  // constrain it to be always greater than 20,
+  // otherwise it won't work in some browsers and systems
+  // reference: https://stackoverflow.com/questions/64473278/gif-frame-duration-seems-slower-than-expected
+  gifFrameDelay = gifFrameDelay < 20 ? 20 : gifFrameDelay;
+
+  // check the mode we are in and how many frames
+  // that duration translates to
+  const nFrames = units === 'seconds' ? duration * _frameRate : duration;
+  const nFramesDelay = units === 'seconds' ? delay * _frameRate : delay;
+  const totalNumberOfFrames = nFrames + nFramesDelay;
+
+  // initialize variables for the frames processing
+  let frameIterator = nFramesDelay;
+  frameCount = frameIterator;
```

Comment:
I am wondering if it is good to set the system variable from here? and also could you please elaborate why it is necessary ?

Reply:
`frameCount` is a variable exposed by the p5.js library upon instancing. Accesing `frameCount` returns the index of the frame we are currently in, and we can overwrite it in order to _travel_ to a given frame.

This is part of the delay feature. We calculate how many frames we should wait for before recording. Imagine setting `delay = 3`, which makes it `3 * 60 = 180` (assuming a frameRate of 60). We have to wait until frame 180 to start recording. Cool, simply set `frameCount = 180` and then keep on as usual. 

This also makes sure that the user, if `delay = 0`, can record their sketch from the very beginning without extra hassle, since this would effectively set `frameCount = 0`, even if the animation has already started.

Reply:
Btw, could this be related to the sketch you mentioned earlier that doesn't work correctly when recording? If a sketch depends partially on `frameCount` but also has some updates that don't (e.g. a physics simulation that updates each `draw()`), I wonder if rewinding `frameCount` might put the sketch into an unexpected state. Maybe there should be a way to keep counting from the current state? e.g. passing `undefined` instead of a number?

Reply:
Yes, it probably will do that. But for sketches not based on that, I cannot think of a way to somehow attach their time component to something. 

I don't know how other systems approach this, but I thought at least we could have support for those who use `frameCount` as a way of dealing with time, since it is a common practice in shader languages. I myself find using it a lot, and also learned about it in some Coding Train video, which may increase the chance of people using it.

This does not solve the problem at all, of course, but at least we are considering a subset of the users for this matter.

I don't know if implementing a robust solution for every sketch is very difficult, but being already familiar with the codebase, I can work on improving it over the next year independently of GSoC or other events, if I can find a way to do it!

Reply:
For sure! I think just supporting frameCount for now is totally fine, as long as we leave space in the API to support other things in the future without breaking older code. So the future uses to consider are:
- sketches that rely on state updates in `draw()`: I did some tests, it seems like doing state updates in `draw()` works fine for gif recording! so this is handled already
- sketches that rely on `millis()`: that can be done later if we provide a mock implementation of `millis`, which won't affect any of the code you've already written, so everything's good here for this PR!
- sketches that mix `frameCount` with state updates (or `millis()`): the only issue is that rewinding will bring `frameCount` out of sync with the others. If we don't rewind, then they're in sync. We don't have to support no rewinding in this PR, e.g. in the future we could support it by supplying `delay: undefined` and removing the check to ensure it's a number. If that looks ok, then everything here is good without changes 🙂  I just want to make sure we're aware of it in case there's a different API we want for it in the future